// @esbuild-plugins/tsconfig-paths/esm/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import fs from 'fs';
import { nodeModuleNameResolver, sys } from 'typescript';
import findUp from 'find-up';
import stripJsonComments from 'strip-json-comments';
const NAME = 'tsconfig-paths';
const debug = require('debug')(NAME);
export function TsconfigPathsPlugin({ name = NAME, absolute = true, onResolved, tsconfig, }) {
    debug('setup');
    const compilerOptions = loadCompilerOptions(tsconfig);
    return {
        name,
        setup: function setup({ onResolve }) {
            onResolve({ filter: /.*/ }, (args) => __awaiter(this, void 0, void 0, function* () {
                const hasMatchingPath = Object.keys((compilerOptions === null || compilerOptions === void 0 ? void 0 : compilerOptions.paths) || {}).some((path) => new RegExp(path.replace('*', '\\w*')).test(args.path));
                if (!hasMatchingPath) {
                    return null;
                }
                const { resolvedModule } = nodeModuleNameResolver(args.path, args.importer, compilerOptions || {}, sys);
                if (!resolvedModule) {
                    return null;
                }
                const { resolvedFileName } = resolvedModule;
                if (!resolvedFileName || resolvedFileName.endsWith('.d.ts')) {
                    return null;
                }
                let resolved = absolute
                    ? sys.resolvePath(resolvedFileName)
                    : resolvedFileName;
                if (onResolved) {
                    onResolved(resolved);
                }
                return {
                    path: resolved,
                };
            }));
        },
    };
}
function loadJSON(p) {
    try {
        let data = fs.readFileSync(p).toString();
        data = stripJsonComments(data);
        return JSON.parse(data);
    }
    catch (e) {
        throw new Error(`Cannot load json for '${p}'`);
    }
}
function loadCompilerOptions(tsconfig) {
    if (!tsconfig) {
        const configPath = findUp.sync(['tsconfig.json', 'jsconfig.json']);
        if (configPath) {
            const config = loadJSON(configPath);
            return config['compilerOptions'] || {};
        }
    }
    if (typeof tsconfig === 'string') {
        if (fs.existsSync(tsconfig)) {
            const config = loadJSON(tsconfig);
            return config['compilerOptions'] || {};
        }
    }
    if (tsconfig && tsconfig['compilerOptions']) {
        return tsconfig['compilerOptions'];
    }
    return {};
}
export default TsconfigPathsPlugin;
