// @esbuild-plugins/tsconfig-paths/esm/index.test.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { build } from 'esbuild';
import { writeFiles } from 'test-support';
import Plugin from '.';
import slash from 'slash';
import path from 'path';
require('debug').enable(require('../package.json').name);
test('works', () => __awaiter(void 0, void 0, void 0, function* () {
    const { unlink, base, paths: [ENTRY, UTILS], } = yield writeFiles({
        'entry.ts': `import {x} from '@custom'; console.log(x);`,
        'utils.ts': `import mod from 'mod'; export const x = mod('x');`,
        'node_modules/mod/index.js': 'export default () => {}',
    });
    let called = 0;
    let resolved = [];
    const tsconfig = {
        baseUrl: '.',
        compilerOptions: { paths: { '@custom': [UTILS] } },
    };
    const res = yield build({
        entryPoints: [ENTRY],
        write: false,
        format: 'esm',
        target: 'es2017',
        bundle: true,
        plugins: [
            Plugin({
                tsconfig,
                onResolved: (p) => {
                    called++;
                    resolved.push(p);
                },
            }),
        ],
    });
    expect(called).toBe(1);
    const expected = ['utils.ts'];
    expect(resolved.map((x) => slash(path.relative(base, x)))).toEqual(expected);
    unlink();
}));
